### PR TITLE
[SPARK-18099][YARN] Fail if same files added to distributed cache for --files and --archives

### PR DESCRIPTION
**Pull request by kishorvpatil**
## What changes were proposed in this pull request?

During spark-submit, if yarn dist cache is instructed to add same file under --files and --archives, This code change ensures the spark yarn distributed cache behaviour is retained i.e. to warn and fail if same files is mentioned in both --files and --archives.
## How was this patch tested?

Manually tested: 
1. if same jar is mentioned in --jars and --files it will continue to submit the job. 
- basically functionality [SPARK-14423] #12203 is unchanged 
  1. if same file is mentioned in --files and --archives it will fail to submit the job.

Please review https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark before opening a pull request.

… under archives and files

### CODE REVIEW

**Comment by tgravescs** [comment]
@kishorvpatil  please look at the test failure

**Comment by tgravescs** [comment]
Jenkins, test this please

**Comment by tgravescs** [comment]
Jenkins, add to whitelist

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -598,8 +598,12 @@ private[spark] class Client(
...
       flist.foreach { file =>
         val (_, localizedPath) = distribute(file, resType = resType)
-        if (addToClasspath && localizedPath != null) {
-          cachedSecondaryJarLinks += localizedPath
+        if (addToClasspath) {
+          if (localizedPath != null) {
+             cachedSecondaryJarLinks += localizedPath
+          }
+        } else {
+          require(localizedPath !=null)
```
Comment (tgravescs) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: add space after !=

Comment (tgravescs) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Lets change the error to illegal argument exception. 
Also lets comment this to indicate jars are ok due to spark 2.0 jar install, everything else shouldn't have multiple of same jar/file/archive.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 2]
```diff
@@ -598,8 +598,16 @@ private[spark] class Client(
...
-        if (addToClasspath && localizedPath != null) {
-          cachedSecondaryJarLinks += localizedPath
+        if (addToClasspath) {
+          if (localizedPath != null) {
+            cachedSecondaryJarLinks += localizedPath
+          }
+        } else {
+          if (localizedPath != null) {
+            throw new IllegalArgumentException(s"Attempt to add ($file) multiple times. " +
+              "Please check the values of 'spark.yarn.dist.files' and/or " +
```
Comment (tgravescs) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: can you remove the part about check values of those specific configs because there are multiple ways for these to specified (configs or --files, --jars, etc).   Perhaps just say please check the values you specified for uploading files,jars, and archives to make sure one isn't specified multiple times..

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 3]
```diff
@@ -598,8 +598,15 @@ private[spark] class Client(
     ).foreach { case (flist, resType, addToClasspath) =>
       flist.foreach { file =>
         val (_, localizedPath) = distribute(file, resType = resType)
-        if (addToClasspath && localizedPath != null) {
-          cachedSecondaryJarLinks += localizedPath
+        if (addToClasspath) {
```
Comment (tgravescs) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: can you add a comment here explaining what exactly thi sis doing to help explain and keep from breaking in future.

Also can you add another unit test to cover this case.

File: yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala [hunk 4]
```diff
@@ -282,6 +282,37 @@ class ClientSuite extends SparkFunSuite with Matchers with BeforeAndAfterAll
...
 
+  test("distribute archive multiple times") {
+    val libs = Utils.createTempDir()
+    val jarsDir = new File(libs, "jars")
+    assert(jarsDir.mkdir())
+    new FileOutputStream(new File(libs, "RELEASE")).close()
+    val userLib1 = Utils.createTempDir()
+    val userLib2 = Utils.createTempDir()
+
+    val jar1 = TestUtils.createJarWithFiles(Map(), jarsDir)
```
Comment (tgravescs) [inline yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala]: not used anywhere

File: yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala [hunk 5]
```diff
@@ -282,6 +282,37 @@ class ClientSuite extends SparkFunSuite with Matchers with BeforeAndAfterAll
     }
   }
 
+  test("distribute archive multiple times") {
+    val libs = Utils.createTempDir()
+    val jarsDir = new File(libs, "jars")
+    assert(jarsDir.mkdir())
```
Comment (tgravescs) [inline yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala]: I don't see jarsDir being used anywhere either

File: yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala [hunk 6]
```diff
@@ -282,6 +282,37 @@ class ClientSuite extends SparkFunSuite with Matchers with BeforeAndAfterAll
...
+    val jar1 = TestUtils.createJarWithFiles(Map(), jarsDir)
+    val jar2 = TestUtils.createJarWithFiles(Map(), userLib1)
+    // Copy jar2 to jar3 with same name
+    val jar3 = {
+      val target = new File(userLib2, new File(jar2.toURI).getName)
+      val input = new FileInputStream(jar2.getPath)
+      val output = new FileOutputStream(target)
+      Utils.copyStream(input, output, closeStreams = true)
+      target.toURI.toURL
+    }
```
Comment (tgravescs) [inline yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala]: If we can cleanup the variables names above I think it would help a lot, the test is confusing.  I know you just copy and pasted but would be nice to clean up.
Also can we have 3 tests or 3 asserts,  

- one for same file in --files
- one for same file in --archives
- one for same file in --files and --archives

**Comment by tgravescs** [comment]
+1

**Comment by ueshin** [comment]
@kishorvpatil @tgravescs It seems this pr is breaking functionalities of `--files` or `--archives`.
Using `--files` or `--archives` with files which are not included to `--jars` doesn't work.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 7]
```diff
@@ -600,10 +600,14 @@ private[spark] class Client(
         val (_, localizedPath) = distribute(file, resType = resType)
         if (addToClasspath) {
           if (localizedPath != null) {
-             cachedSecondaryJarLinks += localizedPath
+            cachedSecondaryJarLinks += localizedPath
           }
         } else {
-          require(localizedPath !=null)
+          if (localizedPath != null) {
```
Comment (ueshin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I guess here is `localizedPath == null` ?

**Comment by tgravescs** [comment]
thanks for pointing this out

**Comment by tgravescs** [comment]
SPARK-18357 filed to fix

**Comment by kishorvpatil** [comment]
@ueshin Sorry about this. The patch is available in https://github.com/apache/spark/pull/15810.

**Comment by ueshin** [comment]
@kishorvpatil Thank you for fixing this!
